migration/src/main/scala/akka/persistence/r2dbc/migration/MigrationTool.scala

```diff
@@ -345,4 +369,113 @@ class MigrationTool(system: ActorSystem[_]) {
+          persistenceIds,
+          states)
+        s
+      case f @ Failure(exc) =>
+        log.error("Migration failed.", exc)
+        f
+    }
+  }
+
+  /**
+   * Migrate Durable State for a single persistence id.
+   */
+  def migrateDurableState(persistenceId: String): Future[Int] = {
+    for {
+      _ <- createProgressTable
+      currentProgress <- migrationDao.currentProgress(persistenceId)
+      stateCount <- migrateDurableState(persistenceId, currentProgress)
+    } yield stateCount
+  }
+
+  private def migrateDurableState(persistenceId: String, currentProgress: Option[CurrentProgress]): Future[Int] = {
+    val progressRevision = currentProgress.map(_.durableStateRevision).getOrElse(0L)
+    loadSourceDurableState(persistenceId, progressRevision + 1).flatMap {
+      case None => Future.successful(0)
+      case Some(selectedDurableState) =>
+        for {
+          revision <- {
+            val serializedRow = serializedDurableStateRow(selectedDurableState)
+            targetDurableStateDao
+              .upsertState(serializedRow, selectedDurableState.value, changeEvent = None)
+              .map(_ => selectedDurableState.revision)(ExecutionContexts.parasitic)
+          }
+          _ <- migrationDao.updateDurableStateProgress(persistenceId, revision)
+        } yield 1
+    }
+  }
+
+  private lazy val checkAssertSingleWriter: Unit = {
+    if (targetR2dbcSettings.durableStateAssertSingleWriter) {
+      throw new IllegalArgumentException(
+        "When running the MigrationTool the " +
+        "`akka.persistence.r2dbc.state.assert-single-writer` configuration must be set to off.")
+    }
+  }
+
+  private def serializedDurableStateRow(selectedDurableState: SelectedDurableState): SerializedStateRow = {
+    val stateAnyRef = selectedDurableState.value.asInstanceOf[AnyRef]
+    val serializedState = serialization.serialize(stateAnyRef).get
+    val stateSerializer = serialization.findSerializerFor(stateAnyRef)
+    val stateManifest = Serializers.manifestFor(stateSerializer, stateAnyRef)
+
+    // not possible to preserve timestamp
+    val timestamp = Instant.now()
+
+    val serializedRow = SerializedStateRow(
+      selectedDurableState.persistenceId,
+      selectedDurableState.revision,
+      timestamp,
+      timestamp,
+      Some(serializedState),
+      stateSerializer.identifier,
+      stateManifest,
+      tags = Set.empty // not possible to preserve tags
+    )
+    serializedRow
+  }
+
+  private def loadSourceDurableState(persistenceId: String, minRevision: Long): Future[Option[SelectedDurableState]] = {
+    if (sourceDurableStatePluginId == "")
+      Future.successful(None)
+    else
+      sourceDurableStateStore
+        .getObject(persistenceId)
+        .map {
+          case GetObjectResult(Some(value), revision) if revision >= minRevision =>
+            Some(SelectedDurableState(persistenceId, revision, value))
+          case _ => None
```

Comment:
Same here, if you pass it nonexistant ids, shouldn't that rather be a failure, or at least a logged warning?

Reply:
since the migration tool is designed to be run while the source system could still be running, it is possible that it has been deleted